File: pkg/listener/src/soundStack.ts
import type { Segment } from "./types.ts"

// Time-orders per-user transcript segments into a single stream. Each `add`
// pushes one user's segment array as a "stack"; `drain` repeatedly emits the
// globally-earliest remaining segment (by `start`), tagging it with its user.
//
// Faithful TS port of the original sound_stack.py (pop-from-front merge), so the
// resulting script.json ordering matches the Python pipeline.
export class SoundStack {
  private readonly sounds: Record<string, Segment[][]> = {}

  add(user: string, segments: Segment[]): void {
    const stack = this.sounds[user] ?? []
    stack.push(segments)
    this.sounds[user] = stack
  }

  next(): Segment | null {
    let lowestUser = ""
    let lowestIdx = -1
    let lowestStart = Number.POSITIVE_INFINITY

    for (const user of Object.keys(this.sounds)) {
      const stacks = this.sounds[user]!
      for (let idx = 0; idx < stacks.length; idx++) {
        const val = stacks[idx]!
        if (val.length === 0) continue
        const start = val[0]!.start ?? Number.POSITIVE_INFINITY
        if (start < lowestStart) {
          lowestUser = user
          lowestIdx = idx
          lowestStart = start
        }
      }
    }

    if (lowestUser === "") return null

    const stk = this.sounds[lowestUser]![lowestIdx]!
    const res = stk.shift()! // pop(0)
    res.user = lowestUser
    return res
  }

  drain(): Segment[] {
    const res: Segment[] = []
    for (;;) {
      const n = this.next()
      if (n === null) break
      res.push(n)
    }
    return res
  }
}

File: pkg/listener/python/sound_stack.py
from typing import Any


class SoundStack:
    sounds: dict[str, list[list[dict[str, Any]]]]

    def __init__(self) -> None:
        self.sounds = {}

    def add(self, user: str, var: list[dict[str, Any]]):
        stack = self.sounds.get(user, [])
        stack.append(var)

        self.sounds[user] = stack

    def next(self) -> dict[str, Any] | None:
        lowest_username = ""
        lowest_idx = -1
        lowest_start = 9999999999

        for username in self.sounds.keys():
            for idx, val in enumerate(self.sounds[username]):
                if len(val) == 0:
                    continue

                start = val[0].get("start", 9999999999)
                if start < lowest_start:
                    lowest_username = username
                    lowest_idx = idx
                    lowest_start = start

        if lowest_username != "":
            stk = self.sounds[lowest_username][lowest_idx]
            res = stk.pop(0)
            self.sounds[lowest_username][lowest_idx] = stk

            res["user"] = lowest_username
            return res

        return None

    def drain(self) -> list[dict[str, Any]]:
        res = []
        while True:
            n = self.next()
            if n is None:
                break

            res.append(n)

        return res
